### PR TITLE
[MIG][16.0] stock_reserve_rule

### PR DESCRIPTION
**Pull request by rousseldenis**
Depends on:
* https://github.com/OCA/product-attribute/pull/1215

### CODE REVIEW

File: stock_reserve_rule/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    "name": "Stock Reservation Rules",
+    "summary": "Configure reservation rules by location",
+    "version": "16.0.1.0.0",
+    "author": "Camptocamp, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/stock-logistics-warehouse",
+    "category": "Stock Management",
+    "depends": [
+        "stock",
+        "stock_helper",
+    ],
```
Comment (jbaudoux) [inline stock_reserve_rule/__manifest__.py:14]:
```suggestion
        "product_packaging_level"
    ],
```

File: stock_reserve_rule/models/stock_reserve_rule.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        "(FIFO, FEFO, ...).\n"
+        "Empty Bins: take goods from a location only if the bin is"
+        " empty afterwards.\n"
+        "Full Packaging: take goods from a location only if the location "
+        "quantity matches a packaging quantity (do not open boxes).",
+    )
+
+    packaging_type_ids = fields.Many2many(
+        comodel_name="stock.package.type",
+        help="Optional packaging when using 'Full Packaging'.\n"
```
Comment (jbaudoux) [inline stock_reserve_rule/models/stock_reserve_rule.py]:
```suggestion
    packaging_level_ids = fields.Many2many(
        comodel_name="product.packaging.level",
        help="Optional packaging level when using 'Full Packaging'.\n"
```

File: stock_reserve_rule/models/stock_reserve_rule.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+    def _apply_strategy_packaging(self, quants):
+        need = yield
+        # Group by location (in this removal strategies, we want to consider
+        # the total quantity held in a location).
+        quants_per_bin = quants._group_by_location()
+
+        product = fields.first(quants).product_id
+
+        packaging_type_filter = self.packaging_type_ids
```
Comment (jbaudoux) [inline stock_reserve_rule/models/stock_reserve_rule.py]:
```suggestion
        packaging_type_filter = self.packaging_level_ids
```

File: stock_reserve_rule/models/stock_reserve_rule.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+        # we'll walk the packagings from largest to smallest to have the
+        # largest containers as possible (1 pallet rather than 10 boxes)
+        packaging_quantities = sorted(
+            product.packaging_ids.filtered(
+                lambda packaging: (
+                    packaging.qty > 0
+                    and (
+                        packaging.package_type_id in packaging_type_filter
```
Comment (jbaudoux) [inline stock_reserve_rule/models/stock_reserve_rule.py]:
```suggestion
                        packaging.packaging_level_id in packaging_type_filter
```

File: stock_reserve_rule/views/stock_reserve_rule_views.xml [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                            <form string="Removal Rule">
+                                <group>
+                                    <field name="name" />
+                                    <field
+                                        name="location_id"
+                                        domain="[('id', 'child_of', parent.location_id)]"
+                                    />
+                                    <field name="removal_strategy" />
+                                    <field
+                                        name="packaging_type_ids"
```
Comment (jbaudoux) [inline stock_reserve_rule/views/stock_reserve_rule_views.xml]:
```suggestion
                                        name="packaging_level_ids"
```

**Review by jbaudoux — CHANGES_REQUESTED** [review]
The v14 package type is replaced by package level in v16

**Review by jbaudoux — APPROVED** [review]

**Review by hparfr — APPROVED** [review]
Functionnal review, LGTM thanks

**Comment by hparfr** [comment]
@rousseldenis can you please remove the last commit `[[DONT MERGE] test-requirements.txt]`

**Review by AungKoKoLin1997 — APPROVED** [review]
LGTM.

File: stock_reserve_rule/models/stock_move.py [hunk 6]
```diff
@@ -98,6 +98,14 @@ def _update_reserved_quantity(
                             strict=strict,
                         )
                         still_need -= taken_in_loc
+                        # We should break between quants if original needs is fulfilled
+                        need_zero = (
+                            float_compare(still_need, 0, precision_rounding=rounding)
+                            != 1
+                        )
+                        if need_zero:
```
Comment (lmignon) [inline stock_reserve_rule/models/stock_move.py]: @rousseldenis 
```python
if  float_is_zero(still_need, precision_rounding=rounding):
```

**Comment by dreispt** [comment]
@rousseldenis Does this have a dependency waiting to be merged?

**Comment by jbaudoux** [comment]
@rousseldenis The dependency has been merged. Can you drop the related commit so that we can merge this one?

**Comment by jbaudoux** [comment]
> @rousseldenis The dependency has been merged. Can you drop the related commit so that we can merge this one?

@rousseldenis The dependency has been merged. Can you drop the related commit so that we can merge this one?

**Pull request merged**